### PR TITLE
Implement system.dns_cache table

### LINKED ISSUE #42690: Add system.dns_cache table
**Use case**

For testing and debug purposes to be able to see internal data structures.

**Describe the solution you'd like**

- To see at least `hostname`, `ipv4` or `ipv6`.
- Additionally can be useful to have time of resolution and when record will expire.

### PR DESCRIPTION
**Pull request by allmazz**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Implemented system.dns_cache table, which can be useful for debugging DNS issues

Closes https://github.com/ClickHouse/ClickHouse/issues/42690

Example usage:  `SELECT * FROM system.dns_cache;`
| hostname | ip\_address | ip\_family | cached\_at |
| :--- | :--- | :--- | :--- |
| localhost | ::1 | IPv6 | 2024-02-11 16:08:16 |
| localhost | 127.0.0.1 | IPv4 | 2024-02-11 16:08:16 |

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit e2ffbf00183532724d14f0337f8d457e75c1c1c7 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59856/e2ffbf00183532724d14f0337f8d457e75c1c1c7/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59856/e2ffbf00183532724d14f0337f8d457e75c1c1c7/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by alexey-milovidov** [comment]
Please sign a CLA.

**Comment by allmazz** [comment]
> Please sign a CLA.

Done

File: tests/queries/0_stateless/02981_system_dns_cache_table.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+select hostname, ip_address, ip_family, (isNotNull(cached_at) and cached_at > '1970-01-01 00:00:00') from system.dns_cache
```
Comment (pufit) [inline tests/queries/0_stateless/02981_system_dns_cache_table.sql]: `isNotNull(cached_at)` unnecessary

Comment (pufit) [reply tests/queries/0_stateless/02981_system_dns_cache_table.sql]: We use the same ClickHouse server for multiple stateless tests. There will likely be other cache entries at this point. The best solution is to make a request to `localhost` explicitly and then search for the entries in `system.dns_cache`.

File: src/Storages/System/StorageSystemDNSCache.cpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+                continue;
+
+            reported_elements.insert(HostIPPair(hostname, ip));
+
+            std::string family_str;
+            switch (address.family())
+            {
+                case Poco::Net::AddressFamily::IPv4:
+                    family_str = "IPv4";
+                break;
```
Comment (pufit) [inline src/Storages/System/StorageSystemDNSCache.cpp:47]: For code style consistency, `break` should be on the same column as the previous line.

File: docs/en/operations/system-tables/dns_cache.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+# dns_cache
+
+Contains information about cached DNS records.
+
+Columns:
+
+- `hostname` ([String](../../sql-reference/data-types/string.md)) — cached hostname
+- `ip_address` ([String](../../sql-reference/data-types/string.md)) — ip address for the hostname
+- `family` ([String](../../sql-reference/data-types/string.md)) — family of the ip address: `IPv4`, `IPv6`, or `UNIX_LOCAL`.
+- `cached_at` ([String](../../sql-reference/data-types/datetime.md)) - datetime when the record has cached
```
Comment (pufit) [inline docs/en/operations/system-tables/dns_cache.md:13]:
```suggestion
- `cached_at` ([String](../../sql-reference/data-types/datetime.md)) - datetime when the record was cached
```

File: src/Common/DNSResolver.cpp [hunk 4]
```diff
@@ -438,6 +435,19 @@ void DNSResolver::addToNewAddresses(const Poco::Net::IPAddress & address)
...
 }
 
+std::vector<std::pair<std::string, DNSResolver::CacheEntry>> DNSResolver::cacheEntries() const
+{
+    std::lock_guard lock(impl->drop_mutex);
+    std::vector<std::pair<std::string, DNSResolver::CacheEntry>> entries;
+
+    for (const auto & [key, entry] : impl->cache_host.dump())
+    {
+        entries.emplace_back(key, *entry);
```
Comment (pufit) [inline src/Common/DNSResolver.cpp:445]:
```suggestion
        entries.emplace_back(std::move(key), *entry);
```

File: src/Common/DNSResolver.cpp [hunk 5]
```diff
@@ -438,6 +435,19 @@ void DNSResolver::addToNewAddresses(const Poco::Net::IPAddress & address)
     impl->new_addresses.insert({address, consecutive_failures});
 }
 
+std::vector<std::pair<std::string, DNSResolver::CacheEntry>> DNSResolver::cacheEntries() const
+{
+    std::lock_guard lock(impl->drop_mutex);
+    std::vector<std::pair<std::string, DNSResolver::CacheEntry>> entries;
+
+    for (const auto & [key, entry] : impl->cache_host.dump())
```
Comment (pufit) [inline src/Common/DNSResolver.cpp:443]:
```suggestion
    for (auto & [key, entry] : impl->cache_host.dump())
```

File: tests/queries/0_stateless/02981_system_dns_cache_table.sql (hunk 1 above)
Comment (allmazz) [reply tests/queries/0_stateless/02981_system_dns_cache_table.sql]: > `isNotNull(cached_at)` unnecessary

No, it's a check for caching time saved properly

Comment (allmazz) [reply tests/queries/0_stateless/02981_system_dns_cache_table.sql]: > There will likely be other cache entries at this point.

thx, it's fixed

**Review by pufit — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
The test has failed. Please resubmit.

File: docs/en/operations/system-tables/dns_cache.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+---
+# dns_cache
+
+Contains information about cached DNS records.
+
+Columns:
+
+- `hostname` ([String](../../sql-reference/data-types/string.md)) — cached hostname
+- `ip_address` ([String](../../sql-reference/data-types/string.md)) — ip address for the hostname
+- `family` ([String](../../sql-reference/data-types/string.md)) — family of the ip address: `IPv4`, `IPv6`, or `UNIX_LOCAL`.
```
Comment (rschu1ze) [inline docs/en/operations/system-tables/dns_cache.md:12]: This should really be of type [Enum](https://clickhouse.com/docs/en/sql-reference/data-types/enum).

Alternatively, replace `ip_address` by `ipv4` (of type `Nullable([IP4](https://clickhouse.com/docs/en/sql-reference/data-types/ipv4))`, `ipv6` (of type `Nulalble(IP6)`) and `unix_local` (of type `String`)

**Comment by allmazz** [comment]
@alexey-milovidov new PR https://github.com/ClickHouse/ClickHouse/pull/60257

File: docs/en/operations/system-tables/dns_cache.md (hunk 6 above)
Comment (allmazz) [reply docs/en/operations/system-tables/dns_cache.md:12]: thx, it's done

**Comment by clickgapai** [comment]
Hi @allmazz @pufit — while reviewing this PR I found the following:

- **Coverage gap:** `system.dns_cache` row data (`cached_at`, `ip_family`) never verified by any test — https://github.com/ClickHouse/ClickHouse/pull/103846

Could you review the test PR and add the `can be tested` label if it looks good?
Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**